C/c4Query.cc

```diff
@@ -235,9 +235,23 @@ void C4Query::enableObserver(C4QueryObserverImpl *obs, bool enable) {
 
 void C4Query::liveQuerierUpdated(QueryEnumerator *qe, C4Error err) {
     Retained<C4QueryEnumeratorImpl> c4e = wrapEnumerator(qe);
-    LOCK(_mutex);
-    if (!_bgQuerier)
-        return;
-    for (auto &obs : _observers)
+    set<C4QueryObserverImpl *> observers;
+    {
+        LOCK(_mutex);
+        if (!_bgQuerier) {
+            return;
+        }
+
+        // CBL-2336: Calling notify inside the lock could result
+        // in a deadlock, but on the other hand not calling it
+        // inside the lock could result in the callback coming back
+        // to mutate the collection while we are using it (which, 
+        // coincidentally, is why this deadlocks in the first place).
+        // So to counteract this, make a copy and iterate over that.
+        observers = _observers;
+    }
+
+    for(auto &obs : observers) {
         obs->notify(c4e, err);
```

Comment:
my concern of taking it out of the lock is whether it can be deleted from another thread. The original lock protects this situation. But really not sure. Defer to @snej

Reply:
I'm not sure even the original situation prevents deletion of the object because it's just a pointer to an externally controlled non ref-counted object.

Reply:
I also noticed the non ref-counted thing. The normal path as I can imagine would start from that observer obj is deleted. In the destructor of the observer, it enableObserver(false), which pulls itself from the list. But if it is locked by liveQuerierUpdated(), it has to wait for "updated" method to finish, then pulls itself out. Then returns to destructor. c.f. C4QueryObserverImpl::~C4QueryObserverImpl()